### PR TITLE
Refactor various metric APIs to make distributed execution easier

### PR DESCRIPTION
**Pull request by na--**
## What?

This PR contains parts of https://github.com/grafana/k6/pull/2816 that are pure refactoring of various APIs, to decouple them slightly and make them more flexible and usable for the purposes of implementing distributed execution and test suites. It does not have any breaking changes and can safely be merged now. I've tried to separate the various changes into logical self-contained and small commits, so I suggest reviewing it commit by commit.

## Why?

Moving these changes to a separate PR that can be merged now reduces the scope of the big PR and makes merging other backwards-compatible parts of it possible. For the parts that aren't backwards-compatible and will remain in a separate WIP PR for a long while, it makes the upkeep of the PR (i.e. rebasing it on top of `master` changes) much, much easier.

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have added tests for my changes.
- [x] I have run linter locally (`make ci-like-lint`) and all checks pass.
- [x] I have run tests locally (`make tests`) and all tests pass.
- [x] I have commented on my code, particularly in hard-to-understand areas.

## Related PR(s)/Issue(s)

- https://github.com/grafana/k6/pull/2816
- https://github.com/grafana/k6/issues/140
- https://github.com/grafana/k6/issues/1342

### CODE REVIEW

File: metrics/engine/engine.go [hunk 1]
```diff
@@ -167,10 +159,10 @@ func (me *MetricsEngine) initSubMetricsAndThresholds() error {
 // StartThresholdCalculations spins up a new goroutine to crunch thresholds and
 // returns a callback that will stop the goroutine and finalizes calculations.
 func (me *MetricsEngine) StartThresholdCalculations(
+	ingester *OutputIngester,
```
Comment (codebien) [inline metrics/engine/engine.go:162]: This is already an improvement, in the final picture, I think we should try to reverse the control here. Passing the MetricsEngine as a dependency to the Ingester (the Output). 

Just sharing this to you for collecting thoughts, not really a request for change.

**Review by codebien — APPROVED** [review]
LGTM, thanks for doing it!

**Review by olegbespalov — APPROVED** [review]
:+1:

**Review by oleiade — APPROVED** [review]
🚀

**Pull request merged**